retis/src/core/filters/meta/filter.rs

```diff
@@ -391,75 +413,131 @@ impl FilterMeta {
             | Type::Restrict(_)
             | Type::DeclTag(_)
             | Type::TypeTag(_) => (),
-            _ => bail!("unexpected type ({})", t.name()),
+            _ => bail!(
+                "unexpected type ({}) while walking struct members",
+                t.name()
+            ),
         };
 
         Ok(false)
     }
 
     // Return all comparable and walkable types Ptr, Int, Array, Enum[64],
     // Struct, Union
-    fn next_walkable(btf: &Btf, r#type: Type) -> Result<(u8, Type)> {
+    fn next_walkable(btf: &Btf, r#type: Type, casted: bool) -> Result<(u8, Type)> {
         let btf_type = r#type.as_btf_type();
         let mut ind = 0;
 
         // Return early if r#type is already walkable
-        if Self::check_one_walkable(&r#type, &mut ind)? {
+        if Self::check_one_walkable(&r#type, &mut ind, casted)? {
             return Ok((0, r#type));
+        } else if casted {
+            return Ok((ind, r#type));
         }
 
         let btf_type = btf_type.ok_or_else(|| {
             anyhow!("cannot convert to iterable type while retrieving next walkable")
         })?;
 
         for x in btf.type_iter(btf_type) {
-            if Self::check_one_walkable(&x, &mut ind)? {
+            if Self::check_one_walkable(&x, &mut ind, casted)? {
                 return Ok((ind, x));
             }
         }
 
         bail!("failed to retrieve next walkable object.")
     }
 
+    fn parse_mask(el: &str) -> Result<u64> {
```

Comment:
We should check `mask != 0` at the end of this function, otherwise setting a mask of 0 actually means setting a mask of `~0` (as mask is not used).

Reply:
yes, this fell through the cracks. Harmless but not needed.